numba/core/dispatcher.py

```diff
@@ -732,7 +732,10 @@ def typeof_pyval(self, val):
         try:
             tp = typeof(val, Purpose.argument)
         except ValueError:
-            tp = types.pyobject
+            self.targetctx.refresh()
```

Comment:
This causes the issue in the CUDA simulator, because refreshing the target context is not thread-safe. This isn't usually an issue because it's rarely or never (I'm not sure which :slightly_smiling_face:) refreshed during multithreaded execution.

I'm not sure whether avoiding refreshing from multiple threads, or making refreshing thread-safe is the way to go - perhaps @stuartarchibald or @sklam might have some thoughts about this?

Reply:
Note that applying the following:

```diff
diff --git a/numba/core/dispatcher.py b/numba/core/dispatcher.py
index b93ca3481..40a9d1d90 100644
--- a/numba/core/dispatcher.py
+++ b/numba/core/dispatcher.py
@@ -732,7 +732,8 @@ class _DispatcherBase(_dispatcher.Dispatcher):
         try:
             tp = typeof(val, Purpose.argument)
         except ValueError:
-            self.targetctx.refresh()
+            with global_compiler_lock:
+                self.targetctx.refresh()
             tp = self.typingctx._get_global_type(val)
             if tp is None:
                 tp = types.pyobject
```

fixes the test, but has a huge overhead. Prior to this PR being applied:

```
$ NUMBA_ENABLE_CUDASIM=1 python runtests.py \
    numba.cuda.tests.cudapy.test_random.TestCudaRandomXoroshiro128p.test_uniform_float32

Ran 1 test in 8.774s

OK
```

With this PR and the above patch:

```
$ NUMBA_ENABLE_CUDASIM=1 python runtests.py \
    numba.cuda.tests.cudapy.test_random.TestCudaRandomXoroshiro128p.test_uniform_float32

Ran 1 test in 49.165s

OK
```

Reply:
@gmarkall, hi, thanks again. I compared the calling times of `BaseContext.refresh()` method, the test `test_uniform_float32` called 4319 times, while the new test `test_as_njit_argument` called 175 times. In fact, I never notice this method `refresh` could be called so many times. @stuartarchibald, @sklam any thought about this?

Reply:
> I'd guess it's getting called so many times because the simulator uses a lot of threads, and they all hit the njit function at roughly the same time.

Emmm, it must be fixed or is acceptable?

Reply:
I think it is acceptable if it is just the cudasim running it that many times and regular cuda execution is not greatly affected. @gmarkall, what's your thought?

Reply:
@dlee992, what is driving the need for the `.refresh()`?
@gmarkall, should the lock happen earlier?

Reply:
> should the lock happen earlier?

@sklam, yes, I also think about moving the lock earlier, to avoid holding/releasing the lock too many times. But I haven't figured out where is the good place to move...